sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -617,57 +617,75 @@ object DateTimeUtils {
     (date - localDate.getDayOfMonth) + localDate.lengthOfMonth()
   }
 
-  // Visible for testing.
-  private[sql] val TRUNC_TO_YEAR = 1
-  private[sql] val TRUNC_TO_MONTH = 2
-  private[sql] val TRUNC_TO_QUARTER = 3
-  private[sql] val TRUNC_TO_WEEK = 4
-  private[sql] val TRUNC_TO_DAY = 5
-  private[sql] val TRUNC_TO_HOUR = 6
-  private[sql] val TRUNC_TO_MINUTE = 7
-  private[sql] val TRUNC_TO_SECOND = 8
+  // The constants are visible for testing purpose only.
   private[sql] val TRUNC_INVALID = -1
+  // The levels from TRUNC_TO_MONTH to TRUNC_TO_MILLENNIUM are used in truncations
+  // of DATE and TIMESTAMP values.
+  private[sql] val TRUNC_TO_MONTH = 0
+  private[sql] val TRUNC_TO_YEAR = 1
+  private[sql] val TRUNC_TO_DECADE = 2
+  private[sql] val TRUNC_TO_CENTURY = 3
+  private[sql] val TRUNC_TO_MILLENNIUM = 4
+  private[sql] val MAX_LEVEL_OF_DATE_TRUNC = TRUNC_TO_MILLENNIUM
+  // The levels from TRUNC_TO_MICROSECOND to TRUNC_TO_QUARTER are used in truncations
+  // of TIMESTAMP values only.
+  private[sql] val TRUNC_TO_MICROSECOND = 5
+  private[sql] val TRUNC_TO_MILLISECOND = 6
+  private[sql] val TRUNC_TO_SECOND = 7
+  private[sql] val TRUNC_TO_MINUTE = 8
+  private[sql] val TRUNC_TO_HOUR = 9
+  private[sql] val TRUNC_TO_DAY = 10
+  private[sql] val TRUNC_TO_WEEK = 11
```

Comment:
So, can you not truncate a date to week, quarter? I understand why it makes no sense to truncate to day or shorter if it's only a date.

Reply:
> So, can you not truncate a date to week, quarter?

For now, `trunc()` (the `TruncDate` expression) cannot do that but you can use `date_trunc()` (`TruncTimestamp` expression) which does implicit conversion from `DATE` to `TIMESTAMP`:
```sql
spark-sql> select date_trunc('quarter', DATE'2019-08-05');
2019-07-01 00:00:00
```
We can support those levels for `trunc()` (`DATE` type) as well.

Reply:
I can make the changes in this PR or should I open a separate JIRA for that?

Reply:
Up to your judgment. If it's worth adding, it seems worth adding here, but if there are arguments for not supporting it (i.e. consistency with another DB) then need not be done.

Reply:
I did that in this PR. Please, take a look.